File: json/json-annotation-processor/src/main/java/ru/tinkoff/kora/json/annotation/processor/writer/WriterFieldType.java
package ru.tinkoff.kora.json.annotation.processor.writer;

import jakarta.annotation.Nullable;
import ru.tinkoff.kora.json.annotation.processor.KnownType;

import javax.lang.model.type.TypeMirror;

public sealed interface WriterFieldType {

    enum JsonValueType {
        VALUE,
        NULLABLE,
        UNDEFINED
    }

    @Nullable
    JsonValueType jsonValueType();

    default boolean isJsonNullable() {
        return jsonValueType() != null
               && (jsonValueType() == JsonValueType.VALUE || jsonValueType() == JsonValueType.NULLABLE);
    }

    default boolean isJsonUndefined() {
        return  jsonValueType() != null
                && (jsonValueType() == JsonValueType.VALUE || jsonValueType() == JsonValueType.UNDEFINED);
    }

    TypeMirror typeMirror();

    record KnownWriterFieldType(KnownType.KnownTypesEnum knownType, TypeMirror typeMirror, JsonValueType jsonValueType) implements WriterFieldType {}

    record UnknownWriterFieldType(TypeMirror typeMirror, JsonValueType jsonValueType) implements WriterFieldType {}
}

File: json/json-symbol-processor/src/main/kotlin/ru/tinkoff/kora/json/ksp/writer/WriterFieldType.kt
package ru.tinkoff.kora.json.ksp.writer

import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSTypeReference
import com.squareup.kotlinpoet.TypeName
import ru.tinkoff.kora.json.ksp.KnownType

sealed interface WriterFieldType {

    val typeName: TypeName

    val type: KSType

    val isJsonNullable: Boolean

    data class KnownWriterFieldType(override val type: KSType, override val typeName: TypeName, override val isJsonNullable: Boolean, val knownType: KnownType.KnownTypesEnum) : WriterFieldType

    data class UnknownWriterFieldType(override val type: KSType, override val typeName: TypeName, override val isJsonNullable: Boolean) : WriterFieldType
}
